pkg/provisioners/message.go

```diff
@@ -62,3 +68,49 @@ func headerSet(headers []string) map[string]bool {
 	}
 	return set
 }
+
+// History returns the list of hosts where the message has been into
+func (m *Message) History() []string {
+	if m.Headers == nil {
+		return nil
+	}
+	if h, ok := m.Headers[MessageHistoryHeader]; ok {
+		return decodeMessageHistory(h)
+	}
+	return nil
+}
+
+// AppendToHistory append a new host at the end of the list of hosts of the message history
+func (m *Message) AppendToHistory(host string) {
+	host = cleanupMessageHistoryItem(host)
+	if host == "" {
+		return
+	}
+	history := m.History()
+	if history == nil {
+		history = make([]string, 0)
+	}
+	newHistory := append(history, host)
+	m.SetHistory(newHistory)
+}
+
+// SetHistory sets the message history to the given value
+func (m *Message) SetHistory(history []string) {
+	historyStr := encodeMessageHistory(history)
+	if m.Headers == nil {
+		m.Headers = make(map[string]string)
+	}
+	m.Headers[MessageHistoryHeader] = historyStr
+}
+
+func cleanupMessageHistoryItem(host string) string {
+	return strings.Trim(host, " ")
+}
+
+func encodeMessageHistory(history []string) string {
+	return strings.Join(history, MessageHistorySeparator)
```

Comment:
Add a space here after MessageHistorySeparator to make the string more readable?

Reply:
(If you do that, you may need to switch to Regexp.Split() in decodeMessageHistory.
```golang
var historySplitter = regexp.MustCompile("\\s*" + MessageHistorySeparator + "\\s*")

func decodeMessageHistory(s string) []string {
  return historySplitter.Split(historyStr, -1)
}
```